content/knowledge_base/flare/consensus/avalanche/specification.md

```diff
@@ -0,0 +1,225 @@
+Node `u` answers a query by checking whether each `T'` such that `T' *<— T` is currently preferred among competing transactions `∀T'' ∈ PT'` from conflict set `P` of `T'`.
+If every single ancestor `T'` fulfills this criterion, the transaction is said to be strongly preferred and receives a yes-vote (1).
+A failure on this criterion at any `T'` yields a no-vote (0).
+When `u` accumulates `k` responses, it checks whether there are `α` yes-votes for `T`, and if so grants the chit value (`cT = 1`).
+This process yields a labeling of the DAG with a chit value and associated confidence for each transaction `T`.
+
+Example of (chit,confidence) values:
+
+```text
+T1(1,6)
+├── T2(1,5)
+│    ├── T4(1,2) ── T8(1,1)
+│    │               │
+│    └────────────── T5(1,3)
+│                    └──── T9(1,1)
+│
+└── T3(0,0)
+    ├── T6(0,0)
+    └── T7(0,0)
+```
+
+Conflict set `P` for `T1` has one element `T1`.
+Conflict set `P` for `T1` has one element `T1`.
+Conflict set `P` for `T2` and `T3` has two elements `{T2, T3}`.
+Conflict set `P` for `T5`, `T6` and `T9` has 3 elements `{T6, T7, T9}`.
+
+Similar to Snowball, sampling in Avalanche creates a positive feedback for the preference of a single transaction in its conflict set.
+When a transaction has a larger confidence, its descendants are likely to collect chits in the future compared to others with lower confidence.
+For example `T2` has higher confidence than `T3` and its descendants are more likely to collect chits in the future.
+
+Similar to Bitcoin, Avalanche leaves determining the acceptance point of a transaction to the application.
+Committing a transaction can be performed through a safe early commitment.
+For virtuous transactions, `T` is accepted when it is the only transaction in its conflict set and has a confidence no less than a threshold of `β1`.
+If a virtuous transaction fails to get accepted due to a problem with parents, it could be accepted if reissued with different parents.
+
+### Transaction Entangling
+
+How Avalanche entangles transactions:
+
+```text
+init:
+    `Set T` = Ø set of known transactions (empty set)
+    Q = Ø set of queried transactions (empty set)
+
+generateTx:
+    edges = {T' <— T: T' ∈ parentSelection(`Set T`)}
+    T = Tx(data, edges)
+    onReceiveEvent(T)
+
+onReceiveEvent(t):
+    if T ∉ `Set T` then
+        if PT (conflict set of T) = Ø then
+            PT = {T}, Pt.pref = T
+            PT.last = T, PT.cnt = 0
+        else PT = PT U {T}
+
+        `Set T` = T U {T}, cT = 0
+```
+
+Because transactions that generate and consume the same UTXO do not conflict with each other, any transaction can be reissued with different parents.
+
+### Node Main Loop
+
+Protocol main loop executed by each node:
+
+```text
+AvalancheLoop:
+    while true do
+        find T that satisfies T ∈ `Set T` AND T ∉ Q
+        K = sample(N\u, k)
```

Comment:
This means `u` not being from N set - since this is a pseudo code -> the reference of `u` in the list of peers must not be a member of N (natural numbers) where N is from there the k comes from

Reply:
Aaaah, right. Okay, that makes sense.